docs_src/design/adr/security/0017-consul-security.md

```diff
@@ -0,0 +1,111 @@
+When Consul is enabled as a configuration provider,
+the `configuration.toml` is parsed into individual settings
+and seeded into the Consul key-value store on the first start of a service.
+Configuration reads and writes are then done to Consul if it is specified as the configuration provider,
+otherwise the static `configuration.toml` is used.
+Writes to the `[Writable]` section in Consul trigger per-service callbacks
+notifying the application of the changed data.
+
+Since configuration data can affect the runtime behavior of services,
+compensating controls must be introduced in order to mitigate the risks introduced
+by moving configuration from a static file into to an HTTP-accessible service with mutable state.
+
+The current practice is that Consul is exposed via unencrypted HTTP in anonymous read/write mode
+to all processes and EdgeX services running on the host machine.
+
+## Decision
+
+Consul will be configured with access control list (ACL) functionality enabled,
+and each EdgeX service will utilize a Consul access token to authenticate to Consul.
+Consul access tokens will be requested from the Vault Consul secrets engine
+(to avoid introducing additional bootstrapping secrets).
+
+DNS will be disabled via configuration as it is not used in EdgeX.
+
+**Consul Access Via API Gateway**
+
+Consul can be configured to be exposed via the EdgeX API Gateway
+in order to provide remote configuration capabilities.
+The UI can also be rebased (in theory) via the `ui_content_path`
+[setting](https://www.consul.io/docs/agent/options#ui_config_enabled)
+in order to be compatible with the API Gateway pathing conventions.
+
+Consul supports authentication via the `X-Consul-Token` HTTP header,
+the `Authorization: Bearer` header,
+or via a `?token=` query parameter (insecure).
+The API gateway should be configured to simply pass HTTP requests
+and rely on Consul's authentication mechanisms for both API and UI accesses.
+For API accesses, the `X-Consul-Token` method should be used
+as it will not conflict with API Gateway authentication.
+For UI accesses, the Consul UI will prompt the user for an access token.
+
+
+## Consequences
+
+Full implementation of this ADR will deny Consul access to all existing Consul clients.
+To limit the impacts of the change, deployment will take place in phases:
+
+### Phase 1
+
+- Vault bootstrapper will install Vault Consul secrets engine.
+- Secretstore-setup will create a Vault token for consul secrets engine configuration.
+- Consul will be started with Consul ACLs enabled with persistent agent tokens and a default "allow" policy.
+- Consul bootstrapper will create a bootstrap management token
+  and use the provided Vault token to (re)configure the Consul secrets engine in Vault.
+- Consul bootstrapper will create an agent ("node") policy and token and install it into the agent.
+- Create a global management token (stored in a file) for use in remote configuration.
```

Comment:
How does one get a hold of the token to use during remote access? Just copy from file?

Reply:
Yeah.

Reply:
Edit: With proxy authentication in place, the only reason a administrator would need to manually access this token would be to use the Consul web interface.